File: algorithm/programmers/level_03/베스트앨범.java
import java.util.*;
import java.util.stream.Collectors;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        Map<String, Songs> results = new HashMap<>();

        for (String genre : genres) {
            results.put(
                    genre,
                    new Songs(new ArrayList<>())
            );
        }

        for (int i = 0; i < genres.length; i++) {
            results.get(genres[i]).add(new Song(i, plays[i]));
        }

        return results.values().stream()
                .sorted(Comparator.comparing(Songs::sum).reversed())
                .map(Songs::getSongs)
                .flatMap(Collection::stream)
                .mapToInt(Integer::intValue)
                .toArray();
    }

    private static class Songs {

        private final List<Song> songs;

        public Songs(List<Song> songs) {
            this.songs = songs;
        }

        public void add(Song song) {
            this.songs.add(song);
        }

        public int sum() {
            return songs.stream()
                    .mapToInt(Song::getPlayCount)
                    .sum();
        }

        public List<Integer> getSongs() {
            return songs.stream()
                    .sorted(
                            Comparator.comparing(Song::getPlayCount).reversed()
                                    .thenComparing(Song::getId)
                    )
                    .map(Song::getId)
                    .limit(2)
                    .collect(Collectors.toUnmodifiableList());
        }
    }

    private static class Song {

        private final int id;
        private final int playCount;

        public Song(int id, int playCount) {
            this.id = id;
            this.playCount = playCount;
        }

        public int getId() {
            return id;
        }

        public int getPlayCount() {
            return playCount;
        }
    }
}

File: algorithm/programmers/level_03/베스트앨범.py
def solution(genres, plays):
    answer = []
    dictionary = dict()

    for genre in genres:
        dictionary[genre] = []

    for index, value in enumerate(plays):
        dictionary[genres[index]].append([index, value])

    dictionary = sorted(dictionary.items(), key=lambda x: -sumPlays(x[1]))

    for value in dictionary:
        songs = sorted(value[1], key=lambda x: (-x[1], x[0]))
        for song in songs[:2]:
            answer.append(song[0])

    return answer


def sumPlays(songs):
    result = 0

    for song in songs:
        result += song[1]

    return result
